src/charm.py

```diff
@@ -26,44 +23,77 @@ class GitLabK8sCharm(CharmBase):
+        # something that we observe which expects the state to be initialized.
+        self.framework.observe(self.on.start, self.init_state)
 
-        self.website = HTTPInterfaceProvides(self, 'website')
-        self.mysql = MySQLInterfaceRequires(self, 'mysql')
+        self.website = HTTPServer(self, 'website')
+        self.mysql = MySQLServer(self, 'mysql')
         self.gitlab_image = OCIImageResource(self, 'gitlab_image')
 
+        # TODO: Make sure that handlers / components setting "waiting" or "active" status doesn't clobber "blocked".
+        self.framework.observe(self.on.leader_elected, self.configure_pod)
+        self.framework.observe(self.on.config_changed, self.configure_pod)
+
+        self.framework.observe(self.mysql.on.database_available, self)
+        self.framework.observe(self.mysql.on.database_changed, self.on_database_available)
+        self.framework.observe(self.mysql.on.database_unavailable, self)
+
+        self.framework.observe(self.gitlab_image.on.image_available, self)
+        self.framework.observe(self.gitlab_image.on.image_failed, self)
+
+        # Must come last to ensure it happens after configure_pod.
+        self.framework.observe(self.website.on.new_client, self)
 
-    def on_install(self, event):
+    def init_state(self, event):
         self.state.is_started = False
+        self.state.has_image = False
+        self.state.registry_path = None
+        self.state.registry_user = None
+        self.state.registry_pass = None
+        self.state.database = None
 
-    def on_start(self, event):
-        unit = self.framework.model.unit
-        if not self.gitlab_image.fetch():
-            unit.status = BlockedStatus('Missing or invalid image resource')
-            return
-        if not self.mysql.is_joined:
-            unit.status = BlockedStatus('Missing database')
-        if not self.mysql.is_single:
-            unit.status = BlockedStatus('Too many databases')
-        if not self.mysql.is_ready:
-            unit.status = WaitingStatus('Waiting for database')
+    def on_image_available(self, event):
+        self.state.has_image = True
+        self.state.registry_path = event.registry_path
+        self.state.username = event.username
+        self.state.password = event.password
+        # Optimistically try to configure the pod.
+        self.configure_pod(event)
+
+    def on_image_failed(self, event):
+        self.state.has_image = False
+        self.model.unit.status = event.status
+
+    def on_database_available(self, event):
+        self.state.database = dict(event.database)  # FIXME: Remove need for dict wrapper
+        # Optimistically try to configure the pod.
+        self.configure_pod(event)
+
+    def on_database_unavailable(self, event):
+        self.state.database = None
+        self.model.unit.status = event.status
+
+    def configure_pod(self, event):
+        if not (self.state.has_image and self.state.database):
```

Comment:
I commented in person, but I do feel like having
try:
 db = self.mysql.database()
except:
 ...
try:
 img_info = self.gitlab_image.?
...

Is much clearer to understand what is going on, rather than boolean attributes on self.state that are essentially just encapsulation of that logic. It follows a lot about validating your arguments at the beginning of the function, rather than assuming you are called with correct arguments.

I *really* like having abstraction for self.mysql.database() to encapsulate all the ideas about only having 1 database, having it ready to operate, etc. And having an encapsulated event for 'Mysql.on.database_changed' is reasonable. (I'm not sure yet whether it is _changed, or separate _available vs _unavailable.)
I can understand some of the conceptual purity of trying to focus how much of the model you have to think about into smaller chunks. But given your final output is a rendered config based on all of the available input, it makes sense to have a centralized location that builds up the available input.